src/qengine/state/opencl_multi.cpp

```diff
@@ -0,0 +1,1276 @@
+//////////////////////////////////////////////////////////////////////////////////////
+//
+// (C) Daniel Strano and the Qrack contributors 2017, 2018. All rights reserved.
+//
+// This is a multithreaded, universal quantum register simulation, allowing
+// (nonphysical) register cloning and direct measurement of probability and
+// phase, to leverage what advantages classical emulation of qubits can have.
+//
+// Licensed under the GNU General Public License V3.
+// See LICENSE.md in the project root or https://www.gnu.org/licenses/gpl-3.0.en.html
+// for details.
+
+#include <algorithm>
+#include <future>
+
+#include "oclengine.hpp"
+#include "qengine_opencl_multi.hpp"
+
+namespace Qrack {
+
+#define CMPLX_NORM_LEN 5
+
+QEngineOCLMulti::QEngineOCLMulti(
+    bitLenInt qBitCount, bitCapInt initState, std::shared_ptr<std::default_random_engine> rgp, int deviceCount)
+    : QInterface(qBitCount)
+{
+
+    rand_generator = rgp;
+
+    runningNorm = 1.0;
+
+    clObj = OCLEngine::Instance();
+    if (deviceCount == -1) {
+        deviceCount = clObj->GetDeviceCount();
+    }
+
+    // deviceCount = 64;
+    bitLenInt devPow = log2(deviceCount);
+    maxDeviceOrder = devPow;
+
+    // Maximum of 2^N devices for N qubits:
+    if (qubitCount <= devPow) {
+        devPow = qubitCount - 1;
+    }
+
+    deviceCount = 1 << devPow;
+    subEngineCount = deviceCount;
+
+    subQubitCount = qubitCount - devPow;
+    subMaxQPower = 1 << subQubitCount;
+    subBufferSize = sizeof(complex) * subMaxQPower >> 1;
+
+    if (deviceCount == 1) {
+        substateEngines.push_back(std::make_shared<QEngineOCL>(qubitCount, initState, rgp));
+        substateEngines[0]->EnableNormalize(true);
+        return;
+    }
+
+    bool foundInitState = false;
+    bool partialInit = true;
+    bitCapInt subInitVal;
+
+    for (int i = 0; i < deviceCount; i++) {
+        if ((!foundInitState) && ((subMaxQPower * (i + 1)) > initState)) {
+            subInitVal = initState - (subMaxQPower * i);
+            foundInitState = true;
+            partialInit = false;
+        }
+        substateEngines.push_back(std::make_shared<QEngineOCL>(subQubitCount, subInitVal, rgp, i, partialInit));
+        substateEngines[i]->EnableNormalize(false);
+        subInitVal = 0;
+        partialInit = true;
+    }
+}
+
+void QEngineOCLMulti::ShuffleBuffers(complex* stateVec1, complex* stateVec2)
```

Comment:
This file looks like it could use some code comments on architecture and approach.

Reply:
Added.